### PR TITLE
CDI-671 SPI element related to Instance are missing info about their qualifier

### PR DESCRIPTION
**Pull request by antoinesd**
CDI-671 SPI element related to Instance are missing info about their qualifier
Adding rules regarding qualifier in different programmatic lookup situation

### CODE REVIEW

File: spec/src/main/asciidoc/core/injectionandresolution.asciidoc [hunk 1]
```diff
@@ -649,6 +649,9 @@ PaymentProcessor pp = anyPaymentProcessor.select(qualifier).get().process(paymen
 
 In this example, the returned bean has qualifier `@Synchronous` or `@Asynchronous` depending upon the value of `synchronously`.
 
+Qualifiers specified at `Instance` injection point or via via the `Instance.select()` method are used to perform typesafe resolution.
```
Comment (tremes) [inline spec/src/main/asciidoc/core/injectionandresolution.asciidoc]: Two times `via`

Comment (tremes) [inline spec/src/main/asciidoc/core/injectionandresolution.asciidoc]: Maybe I would use "specified by" but I would unify this in this PR - in the next file there is "specified in". WDYT?

File: spec/src/main/asciidoc/core/injectionandresolution.asciidoc [hunk 2]
```diff
@@ -649,6 +649,9 @@ PaymentProcessor pp = anyPaymentProcessor.select(qualifier).get().process(paymen
...
+If no qualifier is specified at `Instance` injection point or via the `Instance.select()` method the `@Default` qualifier is assumed.
+
```
Comment (mkouba) [inline spec/src/main/asciidoc/core/injectionandresolution.asciidoc]: I find these sentences redundant.

The section "5.6.1. The Instance interface" clearly states that:
> For an injected Instance, the required qualifiers are the qualifiers specified at the injection point.

And "3.8. The default qualifier at injection points" states:
> If an injection point declares no qualifier, the injection point has exactly one qualifier, the default qualifier @Default.

I understand that section "5.6. Programmatic lookup" is not normative, just a general description and examples. Still, I would reword the sentences to reference the above mentioned sections or to be more PaymentProcessor-example-related.

File: spec/src/main/asciidoc/core/spi.asciidoc [hunk 3]
```diff
@@ -258,6 +258,9 @@ A portable extension or other object may obtain a reference to the current conta
 Other methods on `CDI` may be called after the application initialization is completed until the application shutdown starts.
 If methods on `CDI` are called at any other time, non-portable behavior results.
 
+When performing lookup of bean instance with `CDI` the qualifiers specified in the `Instance.select()` method are used to perform typesafe resolution.
```
Comment (mkouba) [inline spec/src/main/asciidoc/core/spi.asciidoc]: For `javax.enterprise.inject.spi.CDI` I would start more generally, e.g.:

`CDI` implements `javax.enterprise.inject.Instance` and therefore might be used to perform programmatic lookup.
If no qualifier is passed to the `Instance.select()` method the `@Default` qualifier is assumed.

File: spec/src/main/asciidoc/core/spi.asciidoc [hunk 4]
```diff
@@ -710,6 +713,9 @@ Instance<Object> createInstance();
 
 Instances of dependent scoped beans obtained with this `Instance` object must be explicitly released by calling `Instance.destroy()` method.
 
+When performing lookup of bean instance with the returned `Instance`, the qualifiers specified in the `Instance.select()` method are used to perform typesafe resolution.
```
Comment (mkouba) [inline spec/src/main/asciidoc/core/spi.asciidoc]: I would leave out this sentence - it's obvious, and add only:
If no qualifier is passed to the `Instance.select()` method the `@Default` qualifier is assumed.

File: spec/src/main/asciidoc/javase/bootstrap_se.asciidoc [hunk 5]
```diff
@@ -101,6 +101,9 @@ If it is called and the container was already shutdown, it throws an `IllegalSta
 * `getBeanManager()` method returns the `BeanManager` (as defined in <<beanmanager>>) for the running container.
 If it is called and the container was already shutdown, it throws an `IllegalStateException`.
 
+When performing lookup of bean instance with `SeContainer` the qualifiers specified in the `Instance.select()` method are used to perform typesafe resolution.
```
Comment (mkouba) [inline spec/src/main/asciidoc/javase/bootstrap_se.asciidoc]: Like with `javax.enterprise.inject.spi.CDI` - this would deserve some general intro.

File: spec/src/main/asciidoc/core/injectionandresolution.asciidoc (hunk 2 above)
Comment (antoinesd) [inline spec/src/main/asciidoc/core/injectionandresolution.asciidoc]: Well, what's not covered is going from `@Default` to `@MyQualifier` and the fact that it removes `@Default` to required qualifiers list.

Comment (mkouba) [inline spec/src/main/asciidoc/core/injectionandresolution.asciidoc]: See also _"If no qualifier is passed to the Instance.select() method the @Default qualifier is assumed."_ sentence below.

**Comment by mkouba** [comment]
@antoinesd I believe the info should be also included in the javadoc.

File: api/src/main/java/javax/enterprise/inject/se/SeContainer.java [hunk 6]
```diff
@@ -17,12 +17,20 @@
...
 import javax.enterprise.inject.spi.BeanManager;
 
 /**
  * Provides access to the current container in Java SE.
  *
+ * <p>
+ * SeContainer implements {@link Instance} and therefore might be used to perform programmatic lookup.
+ * If no qualifier is specified in {@link Instance#select(Class, Annotation...)} method the <code>@Default</code> qualifier is assumed.
+ * </p>
+ * </p>
```
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/se/SeContainer.java]: Superfluous end of paragraph.

File: spec/src/main/asciidoc/core/spi.asciidoc [hunk 7]
```diff
@@ -258,6 +258,9 @@ A portable extension or other object may obtain a reference to the current conta
 Other methods on `CDI` may be called after the application initialization is completed until the application shutdown starts.
 If methods on `CDI` are called at any other time, non-portable behavior results.
 
+CDI implements `javax.enterprise.inject.Instance` and therefore might be used to perform programmatic lookup as defined in <<dynamic_lookup>>.
+If no qualifier is specified in `Instance.select()` method the `@Default` qualifier is assumed.
```
Comment (mkouba) [inline spec/src/main/asciidoc/core/spi.asciidoc]: If no qualifier is passed to `CDI.select()` method the `@Default` qualifier is assumed.

File: api/src/main/java/javax/enterprise/inject/se/SeContainer.java [hunk 8]
```diff
@@ -17,12 +17,19 @@
...
+
 import javax.enterprise.inject.Instance;
 import javax.enterprise.inject.spi.BeanManager;
 
 /**
  * Provides access to the current container in Java SE.
  *
+ * <p>
+ * SeContainer implements {@link Instance} and therefore might be used to perform programmatic lookup.
+ * If no qualifier is specified in {@link Instance#select(Class, Annotation...)} method the <code>@Default</code> qualifier is assumed.
```
Comment (manovotn) [inline api/src/main/java/javax/enterprise/inject/se/SeContainer.java]: Add comma:
> ... method, the ...

Same thing in your changes (spec & javadoc) below.

File: spec/src/main/asciidoc/javase/bootstrap_se.asciidoc [hunk 9]
```diff
@@ -101,6 +101,9 @@ If it is called and the container was already shutdown, it throws an `IllegalSta
 * `getBeanManager()` method returns the `BeanManager` (as defined in <<beanmanager>>) for the running container.
 If it is called and the container was already shutdown, it throws an `IllegalStateException`.
 
+SeContainer implements `javax.enterprise.inject.Instance` and therefore might be used to perform programmatic lookup as defined in <<dynamic_lookup>>.
+If no qualifier is specified in `Instance.select()` method the `@Default` qualifier is assumed.
```
Comment (mkouba) [inline spec/src/main/asciidoc/javase/bootstrap_se.asciidoc]: If no qualifier is passed to `SeContainer.select()` method the `@Default` qualifier is assumed.

File: api/src/main/java/javax/enterprise/inject/se/SeContainer.java (hunk 8 above)
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/se/SeContainer.java]: I would rather use "If no qualifier is passed to" in all places... and `{@link #select(Class, Annotation...)}`.

File: spec/src/main/asciidoc/core/spi.asciidoc (hunk 7 above)
Comment (tremes) [inline spec/src/main/asciidoc/core/spi.asciidoc]: Wouldn't be more clear to explicitly mention `CDI.current().select()` instead of `Instance.select()`? It's just minor....

File: api/src/main/java/javax/enterprise/inject/se/SeContainer.java (hunk 8 above)
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/se/SeContainer.java]: Also there are three select() methods which accept qualifiers.

**Comment by mkouba** [comment]
@manovotn `@Inject Instance<AnyClass>` is already covered by the spec, see also "3.8. The default qualifier at injection points" and "5.6.1. The Instance interface".

File: spec/src/main/asciidoc/core/spi.asciidoc (hunk 7 above)
Comment (antoinesd) [inline spec/src/main/asciidoc/core/spi.asciidoc]: we go for CDI.select()

**Comment by mkouba** [comment]
Looks good.

**Comment by manovotn** [comment]
+1

**Comment by tremes** [comment]
LGTM
